_posts/2020-08-05-Mocks-and-Spies-in-Minitest.md

```diff
@@ -0,0 +1,78 @@
+---
+layout: post
+title: "Mocks & Spies in Minitest"
+author: podrezo
+date: 2020-08-05
+excerpt: "For one of our smaller ruby projects we didn't want to bring in all of rspec, so we decided to go with minitest as our unit test runner and assertion library. Doing mocks and spies in minitest proved a little bit challenging to get the hang of initially but it's actually quite easy. This article will explain how to do it."
+---
+For one of our smaller ruby projects we didn't want to bring in all of rspec, so we decided to go with minitest as our unit test runner and assertion library. Doing mocks and spies in minitest proved a little bit challenging to get the hang of initially but it's actually quite easy. This article will explain how to do it.
+
+First, let's consider two classes - one helper class which we will be mocking out, and the other is the class under test.
+
+
+```ruby
+class MyHelperClass
+  def upcase_instance(value)
+    # value.upcase
+    raise "Fail the test if the real thing is called"
+  end
+
+  def self.upcase_static(value)
+    # value.upcase
+    raise "Fail the test if the real thing is called"
+  end
+end
+
+class MyClassUnderTest
+  def use_upcase_instance(value)
+    helper_class = MyHelperClass.new
+    helper_class.upcase_instance(value)
+  end
+
+  def use_upcase_static(value)
+    MyHelperClass.upcase_static(value)
+  end
+end
+```
+
+Note that if the real helper class's methods are called the test will fail to illustrate that we're doing the correct thing. In your test spec file you wil need to bring in the necessary `require`s for minitest:
+
+```ruby
+require "minitest/autorun"
+require "minitest/mock"
+```
+
+The core aspect of mocking in minitest comes from `Minitest::Mock`'s `expect` method coupled with the `stub` method on whatever you're mocking. The [expect method signature](https://www.rubydoc.info/gems/minitest/Minitest/Mock) looks like this:
+
+```ruby
+expect(name, retval, args = [], &blk) ⇒ Object
+# Expect that method name is called, optionally with args or a blk, and returns retval.
```

Comment:
What is this comment adding here?

In my opinion it could be valuable if it was clear about what is the `retval` and if it called `blk` a `block` to be more explicit. The way it is I feel like it's just teaching ruby to people reading the article, which shouldn't be the objective here.
Does this comment make sense to you?

Reply:
This is from the official docs; it says what each parameter does.  It was copy and paste because; I figured it may be useful because it is not immediately clear if "args" means the args for the method you're mocking or additional arguments to the expect method

Reply:
🤷  up to you